shell: mqtt: use topic levels

jeppenodgaard opened the pull request:
Change topic from <device_id>_rx (and tx) to <device_id>/sh/rx.

This allows use of wildcards. E.g. subscribe to all devices "+/sh/tx".

Level "sh" is added to the topic to make it less generic and prevent potential clashes with other topics.

Commit:
shell: mqtt: use topic levels

Change topic from <device_id>_rx (and tx) to <device_id>/sh/rx.

This allows use of wildcards. E.g. subscribe to all devices "+/sh/tx".

Level "sh" is added to the topic to make it less generic and prevent
potential clashes with other topics.

The topic part after <device_id> is configurable via Kconfig.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

kartben reviewed (changes requested):
I don't mind the change but this will be breaking users so needs to be mentioned in the migration guide.

Bonus points if:
- you consider adding some documentation for the MQTT backend in https://docs.zephyrproject.org/latest/services/shell/index.html#backends (of course this would be a different PR, not asking you to do it here)
- you consider making the topics being used something that can be configured via Kconfig, since things are pretty inflexible at the moment, as your PR basically highlights. I would argue this could/should be done in this very PR since it would effectively make this change non-breaking for folks if you keep the same default value as before.

Thanks!

jeppenodgaard commented:
Thank you for your review and insightful feedback @kartben .

> you consider making the topics being used something that can be configured via Kconfig, since things are pretty inflexible at the moment, as your PR basically highlights. I would argue this could/should be done in this very PR since it would effectively make this change non-breaking for folks if you keep the same default value as before.

I think it is difficult to make something that is backward compatible and still follow MQTT best practice.

I would like to have `SHELL_MQTT_TOPIC_IDENTIFIER` which is default `sh`, `SHELL_MQTT_TOPIC_TX_IDENTIFIER` which is default `tx` and `SHELL_MQTT_TOPIC_RX_IDENTIFIER` which is default `rx`. This approach does not allow using `_` between device identifier and the rest of the topic.

The backward compatible solution would be to make only two Kconfig which would have shared MQTT levels:
- SHELL_MQTT_TOPIC_RX_IDENTIFIER with default `/sh/rx`
- SHELL_MQTT_TOPIC_TX_IDENTIFIER with default `/sh/tx`

Which is less granular.

Semi-related: I think there is a need for a common MQTT topic device identifier in Kconfig somewhere, so everything that uses MQTT has the same identifier.

Commit:
shell: mqtt: select NET_CONNECTION_MANAGER

Replace `select NET_MGMT` and `select NET_MGMT_EVENT` with
`select NET_CONNECTION_MANAGER` to get L4 events.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

kartben reviewed (dismissed):
Nice! Sorry for the delay in coming back to this and thank you for your patience
